### ISSUE TITLE
Y24-422 - As an SSR, I would like to bulk upload study data from a CSV file generated from the SSR's Study Setup Google Form into Sequencescape

### ISSUE THREAD

**Issue by harrietc52** [sanger/sequencescape#4446]
**User story**
As the SSR team, we would like to be able to bulk upload study data from a CSV file generated from our Google Study Setup Form, to reduce the manual load and errors created when Study data is copy and pasted into the Sequencescape study creation page. 

Currently, for HumGen and GenGen, Study data is added to a word document, which it is then copied into Sequencescape from. This is being replaced with a Google Form that outputs a CSV. 

This form would support the data required for both EGA & ENA release strategies. 

This story is part for the [LeGo project](https://github.com/orgs/sanger/projects/16/views/1?filterQuery=label%3ALeGo), but it is not part of the MVP. 

**Who are the primary contacts for this story**
@harrietc52 

**Who is the nominated tester for UAT**
Liz C, Avi

**Acceptance criteria**
To be considered successful the solution must allow:

- [ ] The SSR team is aware of the required Study field names in Sequencscape
- [ ] The SSR team is aware of the Study field data validation in Sequencescsape
- [ ] Any Study fields which exist in the Google form, but not Sequencesacpe, are added to Sequencescapes Study Setup page with the expected type, requirement and validation
- [ ] The SSR team has provided a Google Form CSV output, where each header maps to a Sequencescapes Study fields, and the data validation matches that defined by Sequencescape (e.g. if the Data Release Strategy is EGA, ensure the EGA fields are required)
- [ ] A new button, on the right hand side column, is added to https://<env>.sequencescape.psd.sanger.ac.uk/studies, called "Bulk Study upload from CSV”, this takes the user to the new page
- [ ] A new page is created to handle the creation of Studies via uploading one CSV
- [ ] The CSV will contain one row per Study.
- [ ] Each header in the CSV maps to Study field
- [ ] Each row in the CSV, contains new Study data
- [ ] The validation for each rows data is the same as if the study was manually entered via the Study Setup page
- [ ] The CSV upload can create 1 or more Studies
- [ ] When a field is added/ removed or updated in the Sequencescape Study set up page, minimal work needs to be done to support this update via the CSV upload 
- [ ] The CSV upload will not be successful if:
- missing required header
- unrecognised header
- missing required data
- data fails validation
- duplicate Study Name 
- [ ] If the CSV upload is not successful, no Studies will be created. 
- [ ] If the CSV upload is not successful, User friendly messages are shown for each scenario
- [ ] Each unsuccessful scenario has an example CSV and a unit test
- [ ] A User can still create a Study via the original manual method on the Sequencescape Study set up page
- [ ] A User can still edit a Study fields via the original method
- [ ] It is clear on the Study Bulk Upload page that the upload only creates Studies which do not exist. (e.g. Please remove any rows with a Study name that already exists in Sequencescape)
- [ ] It is clear on the UI SS Study Setup page, and within the code comments, that when any change is made to the Study Setup page, the Stakeholders (SSR team) must be made aware of this change, so that they can update their Google Study Setup Form
- [ ] It is clear in the UI SS Study Setup page code comments, that any edit to this form should also be tested via the CSV creation of the Studies.  
- [ ] It is agreed that when a change is made to the Google Form, the SSR team will put in a Story to PSD so that Sequencescape can be updated

**Dependencies**
This story is blocked by the following dependencies:
- #4443 

**References**
This story has a non-blocking relationship with:
- Research story [Y24-374](https://github.com/sanger/sequencescape/issues/4407)

**Additional context**

Mapping Sequencescape fields with Google Form CSV output:
[PSD Google Sheet with DAA fields](https://docs.google.com/spreadsheets/d/1Ya6YyBuoGNBV-6xFLfGHc1WuLjeVLjSlCOMQC6VCUhQ/edit?usp=sharing)

**Comment by harrietc52** [comment]
Questions (Liz C)
- Where would you like this to appear? e.g. as a Link on the right pane “Bulk Study upload from CSV”, which opens to a new page?
- Do you want to download a template CSV, which would ensure the headers were correct, with data validation, or would you upload the CSV which was exported from Liz C’s Google Form directly?
- - If a template CSV download is required: header should reflect the fields from the Study creation page, with what data validation?
- - If a template CSV is not required, Liz C Google Form CSV export headers should reflect the fields from the Study creation page

**Comment by TWJW-SANGER** [comment]
I think it would be good to consider the behaviour for the following use cases:

- User uploads spreadsheet that fails to be read or with expected column headers
- User uploads spreadsheet - and 1 or more of the studies fail validation. (Do the good ones get created or reject all?)
- User uploads spreadsheet - and 1 or more of the studies already exist. (Do the existing studies get updated or do we reject. Note EDAM data cannot be updated post creation)

For all the cases I would err on the simplest implementation side, unless there is a good reason not to. That is reject all changes if there are any errors. Reject if any study already exists.

**Comment by harrietc52** [comment]
Closing for now as not part of the LeGo MVP
